src_zlib/service/zlib_service.cpp

```diff
@@ -214,6 +234,163 @@ class probe
 
 //------------------------------------------------
 
+
+namespace {
+void* zalloc_impl(
+    void* /* opaque */,
+    unsigned items,
+    unsigned size)
+{
+    try {
+        return ::operator new(items * size);
+    } catch(std::bad_alloc const&) {
+        return Z_NULL;
+    }
+}
+
+void zfree_impl(void* /* opaque */, void* addr)
+{
+    ::operator delete(addr);
+}
+} // namespace
+
+class BOOST_HTTP_PROTO_ZLIB_DECL
+    zlib_filter final : public filter
+{
+private:
+    z_stream stream_;
+
+    void init(bool use_gzip);
+
+public:
+    zlib_filter(bool use_gzip = false);
+    ~zlib_filter();
+
+    zlib_filter(zlib_filter const&) = delete;
+    zlib_filter& operator=(zlib_filter const&) = delete;
+
+    filter::results
+    on_process(
+        buffers::mutable_buffer out,
+        buffers::const_buffer in,
+        bool more) override;
+};
+
+zlib_filter::
+zlib_filter(bool use_gzip)
+{
+    stream_.zalloc = &zalloc_impl;
+    stream_.zfree = &zfree_impl;
+    stream_.opaque = nullptr;
+    init(use_gzip);
+}
+
+zlib_filter::
+~zlib_filter()
+{
+    deflateEnd(&stream_);
+}
+
+void
+zlib_filter::
+init(bool use_gzip)
+{
+    int ret = -1;
+
+    int window_bits = 15;
+    if( use_gzip )
+        window_bits += 16;
+
+    int mem_level = 8;
+
+    ret = deflateInit2(
+        &stream_, Z_DEFAULT_COMPRESSION, Z_DEFLATED,
+        window_bits, mem_level, Z_DEFAULT_STRATEGY);
+
+    if( ret != Z_OK )
+        throw ret;
+
+    stream_.next_out = nullptr;
+    stream_.avail_out = 0;
+
+    stream_.next_in = nullptr;
+    stream_.avail_in = 0;
+}
+
+filter::results
+zlib_filter::
+on_process(
+    buffers::mutable_buffer out,
+    buffers::const_buffer in,
+    bool more)
+{
+    auto& zstream = stream_;
+
+    BOOST_ASSERT(out.size() > 6);
+
+    auto flush = more ? Z_NO_FLUSH : Z_FINISH;
+    int ret = -1;
+    filter::results results;
+
+    while( true )
+    {
+        zstream.next_in =
+            reinterpret_cast<unsigned char*>(
+                const_cast<void*>(in.data()));
+        zstream.avail_in = static_cast<unsigned>(
+            in.size());
+
+        zstream.next_out =
+            reinterpret_cast<unsigned char*>(
+                out.data());
+        zstream.avail_out =
+            static_cast<unsigned>(out.size());
+
+        auto n1 = zstream.avail_in;
+        auto n2 = zstream.avail_out;
+        ret = deflate(&zstream, flush);
+
+        in += (n1 - zstream.avail_in);
+        out += (n2 - zstream.avail_out);
+
+        results.in_bytes += (n1 - zstream.avail_in);
+        results.out_bytes += (n2 - zstream.avail_out);
+
+        auto is_empty = (in.size() == 0);
+
+        if( ret != Z_OK &&
+            ret != Z_BUF_ERROR &&
+            ret != Z_STREAM_END )
+            throw ret;
```

Comment:
this needs to be a formal throw, i.e. from a detail function which has the source_location and all that.